sycl/include/sycl/ext/oneapi/experimental/address_cast.hpp

```diff
@@ -16,29 +16,29 @@ inline namespace _V1 {
 namespace ext {
 namespace oneapi {
 namespace experimental {
+namespace detail {
+using namespace sycl::detail;
+}
 // Shorthands for address space names
 constexpr inline access::address_space global_space = access::address_space::global_space;
 constexpr inline access::address_space local_space = access::address_space::local_space;
 constexpr inline access::address_space private_space = access::address_space::private_space;
 constexpr inline access::address_space generic_space = access::address_space::generic_space;
 
 template <access::address_space Space, typename ElementType>
-multi_ptr<ElementType, Space, access::decorated::no>
+multi_ptr<std::remove_pointer_t<remove_decoration_t<ElementType *>>, Space,
+          access::decorated::no>
 static_address_cast(ElementType *Ptr) {
-  using ret_ty = multi_ptr<ElementType, Space, access::decorated::no>;
+  using ret_ty =
+      multi_ptr<std::remove_pointer_t<remove_decoration_t<ElementType *>>,
+                Space, access::decorated::no>;
 #ifdef __SYCL_DEVICE_ONLY__
-  // TODO: Remove this restriction.
-  static_assert(std::is_same_v<ElementType, remove_decoration_t<ElementType>>,
-                "The extension expect undecorated raw pointers only!");
-  if constexpr (Space == generic_space) {
-    // Undecorated raw pointer is in generic AS already, no extra casts needed.
-    // Note for future, for `OpPtrCastToGeneric`, `Pointer` must point to one of
-    // `Storage Classes` that doesn't include `Generic`, so this will have to
-    // remain a special case even if the restriction above is lifted.
+  constexpr auto SrcAS = detail::deduce_AS<ElementType *>::value;
+  if constexpr (SrcAS == Space || Space == generic_space) {
     return ret_ty(Ptr);
   } else {
-    auto CastPtr = sycl::detail::spirv::GenericCastToPtr<Space>(Ptr);
-    return ret_ty(CastPtr);
+    static_assert(SrcAS == generic_space, "Invalid address cast!");
```

Comment:
I had to read this several times to convince myself that the logic was correct.  I think it would be a lot clearer for future readers with some comments.  It might even be worth splitting the `else` case in two for readability, but I appreciate that's really a question of style.

For example, here's how I wrote it out while trying to make sure I could follow the logic:
```c++
if constexpr (SrcAS == Space || Space == generic_space) {
  // if spaces match or target space is generic, no cast is required
  return ret_ty(Ptr);
}
else if constexpr (SrcAS == generic_space) {
  // if the pointer's address space is unknown, try to cast
  return ret_ty(sycl::detail::spirv::GenericCastToPtr<Space>(Ptr));
}
else /* (SrcAS != Space && SrcAS != generic_space) */ {
  // if we can prove the address spaces don't match, behavior is undefined
  static_assert(SrcAS == Space || SrcAS == generic_space, "Invalid address cast!");
}
```